### PR TITLE
Support non-standard ports for upstreams

### PR DESCRIPTION
**Pull request by drewkerrigan**
Addresses the Gen portion of https://github.com/mesosphere/mesos-dns/issues/459

### CODE REVIEW

**Comment by sargun** [comment]
I think you're going to want to use a separate validation function because validate_ipv4_addrs is used by validate_host_list, which in turn is used by the master list as well. In addition, calculate_mesos_dns_resolvers_str might need to be changed, I'm not sure.

**Comment by cmaloney** [comment]
Yea, most the validate_host_list expect no port (and should hard assert there is no port). This will need it's own validation function which checks with port.

**Comment by drewkerrigan** [comment]
@sargun `calculate_mesos_dns_resolvers_str` should be fine, it just makes a decision based on the length of the list

**Comment by cmaloney** [comment]
@drewkerrigan the code in gen resolvconf sets up /etc/resolv.conf with the "upstream" DNS if spartan isn't up for whatever reason (https://github.com/dcos/dcos/blob/master/packages/spartan/extra/gen_resolvconf.py#L63). `/etc/resolv.conf` doesn't allow custom port resolvers.

Need to have some solution for that (Could maybe exclude non-default-port resolvers, although I could see that causing some problems).

**Comment by drewkerrigan** [comment]
@cmaloney yes that could be a problem. I think skipping is ok as long as it is clearly noted like it is here: https://github.com/dcos/dcos/blob/master/packages/spartan/extra/gen_resolvconf.py#L31

Thoughts?

**Comment by drewkerrigan** [comment]
To make sure it doesn't get lost in the mix, this functionality also relies on https://github.com/mesosphere/mesos-dns/pull/468

File: gen/calc.py [hunk 1]
```diff
@@ -224,6 +224,17 @@ def validate_host_list(host_list):
         return host_list
     assert not any(azure_format_check), "Azure static master list and IP based static master list cannot be mixed. Use "
     "either all Azure IP references or IPv4 addresses."
+
```
Comment (cmaloney) [inline gen/calc.py]: this should be a separate function. Technically duplicate IPs with different ports are fine. I'm not a fan of making a function that radically changes behavior based on a parameter.

File: packages/spartan/extra/gen_resolvconf.py [hunk 2]
```diff
@@ -69,7 +69,20 @@ def check_server(addr):
...
+    fallback_servers = []
+
+    # Resolvconf does not support custom ports, skip if not default
+    for ns in os.environ['RESOLVERS'].split(','):
+        ip, separator, port = ns.rpartition(':')
+        if not separator:
+            fallback_servers.append(ns)
+            continue
+        if port == "53":
+            fallback_servers.append(ip)
```
Comment (cmaloney) [inline packages/spartan/extra/gen_resolvconf.py]: This doesn't seem to match the docs: https://docs.python.org/3.4/library/stdtypes.html#str.rpartition

Comment (drewkerrigan) [inline packages/spartan/extra/gen_resolvconf.py]: @cmaloney Not sure what you mean:

```
>>> host, sep, port = '192.168.1.1'.rpartition(':')
>>> not sep
True
>>> host, sep, port = '192.168.1.1:53'.rpartition(':')
>>> not sep
False
```

Are you just suggesting a stricter check on separator being an empty string or populated?

File: gen/calc.py (hunk 1 above)
Comment (drewkerrigan) [inline gen/calc.py]: The duplicate check happens on the list of hosts after json decode, so duplicate IPs with different ports wouldn't break the assertion as written.

Re: separating the function, I see now that I did break the norm by adding a default argument, will address

File: packages/spartan/extra/gen_resolvconf.py (hunk 2 above)
Comment (cmaloney) [inline packages/spartan/extra/gen_resolvconf.py]:
```
>>> "1:2".rpartition(':')
('1', ':', '2')
>>> "2".rpartition(':')
('', '', '2')
```
In the case wehere sep doesn't exist, the data ends up in port / the last chunk. From the docs:
"If the separator is not found, return a 3-tuple containing two empty strings, followed by the string itself." -- the first two end up empty.

Comment (drewkerrigan) [inline packages/spartan/extra/gen_resolvconf.py]: correct, when sep isn't there, the code adds `ns` to `fallback_servers`. Otherwise it will use `ip`

**Comment by cmaloney** [comment]
@drewkerrigan can you rebase on master and fix conflicts. Things like the validate_host_port_list will run into stuff now (that code was reworked in gen/calc.py)

File: gen/calc.py [hunk 3]
```diff
@@ -227,6 +227,30 @@ def validate_host_list(host_list):
...
 
 
+def validate_host_port_list(host_list):
+    host_list = validate_json_list(host_list)
+    azure_format_check = []
+    ips = []
+    validate_duplicates(host_list)
+    for host in host_list:
+        assert isinstance(host, str), 'Host must be of type string, got {}'.format(type(host))
+        if host.startswith('[[[reference(') and host.endswith(').ipConfigurations[0].properties.privateIPAddress]]]'):  # noqa
```
Comment (sargun) [inline gen/calc.py]: Comment, what is this behavior?

File: gen/calc.py [hunk 4]
```diff
@@ -435,7 +459,7 @@ def validate_os_type(os_type):
         'provider': {
             'onprem': {
                 'default': {
-                    'resolvers': '["8.8.8.8", "8.8.4.4"]',
+                    'resolvers': '["8.8.8.8", "8.8.4.4:53"]',
```
Comment (sargun) [inline gen/calc.py]: Why? Just a matter of example, or testing?

File: packages/spartan/extra/gen_resolvconf.py [hunk 5]
```diff
@@ -69,7 +69,20 @@ def check_server(addr):
...
+            continue
+        print('Skipping DNS server {}: non-default ports are not supported in resolvconf'.format(
```
Comment (sargun) [inline packages/spartan/extra/gen_resolvconf.py]: /etc/resolv.conf -- resolvconf is yet another piece of DNS configuration software

File: gen/calc.py (hunk 4 above)
Comment (drewkerrigan) [inline gen/calc.py]: Yes that was for testing, it will be removed it the next commit.

**Comment by drewkerrigan** [comment]
@cmaloney Conflicts taken care of.
@sargun I've addressed your comments, take another look when you get a chance.

**Comment by cmaloney** [comment]
:ship:

**Comment by cmaloney** [comment]
In hold until https://github.com/mesosphere/mesos-dns/pull/468 makes it into a branch this can point to

**Comment by drewkerrigan** [comment]
@cmaloney mesos-dns version bumped.

**Comment by cmaloney** [comment]
:ship:

**Pull request merged**